futures/tests/io_read_until.rs

```diff
@@ -35,53 +35,14 @@ fn run<F: Future + Unpin>(mut f: F) -> F::Output {
     }
 }
 
-struct MaybePending<'a> {
-    inner: &'a [u8],
-    ready: bool,
-}
-
-impl<'a> MaybePending<'a> {
-    fn new(inner: &'a [u8]) -> Self {
-        Self { inner, ready: false }
-    }
-}
-
-impl AsyncRead for MaybePending<'_> {
-    fn poll_read(self: Pin<&mut Self>, _: &mut Context<'_>, _: &mut [u8])
-        -> Poll<io::Result<usize>>
-    {
-        unimplemented!()
-    }
-}
-
-impl AsyncBufRead for MaybePending<'_> {
-    fn poll_fill_buf<'a>(mut self: Pin<&'a mut Self>, _: &mut Context<'_>)
-        -> Poll<io::Result<&'a [u8]>>
-    {
-        if self.ready {
-            self.ready = false;
-            if self.inner.is_empty() { return Poll::Ready(Ok(&[])) }
-            let len = cmp::min(2, self.inner.len());
-            Poll::Ready(Ok(&self.inner[0..len]))
-        } else {
-            self.ready = true;
-            Poll::Pending
-        }
-    }
-
-    fn consume(mut self: Pin<&mut Self>, amt: usize) {
-        self.inner = &self.inner[amt..];
-    }
-}
-
 #[test]
 fn maybe_pending() {
-    let mut buf = MaybePending::new(b"12");
+    let mut buf = b"12".interleave_pending();
     let mut v = Vec::new();
     assert_eq!(run(buf.read_until(b'3', &mut v)).unwrap(), 2);
     assert_eq!(v, b"12");
 
-    let mut buf = MaybePending::new(b"12333");
+    let mut buf = b"12333".interleave_pending();
```

Comment:
Note that this won't add a pending in between reads that actually return data like the old implementation, it will be like `Pending`, `Ready(b"123")`, `Pending`, `Ready(b"")`.

If you want to test reading data multiple times my plan elsewhere was to use `TryStreamExt::into_async_read`, so something like `stream::iter(vec![b"12", b"33", b"3"]).map(Ok).into_async_read().interleave_pending()`.

Reply:
Thanks for pointing that out. I updated this PR after checking locally that the new test can detect problems like https://github.com/rust-lang-nursery/futures-rs/pull/1566, but the code you suggested is even better. I will open a PR

Reply:
Hmm... `thread 'maybe_pending' panicked at 'Attempted to consume from IntoAsyncRead without chunk'` at
https://github.com/rust-lang-nursery/futures-rs/blob/f4dad6654b65be08bb0b59f2a274ad18d96e436a/futures/tests/io_read_until.rs#L56

This is because `read_until` runs `consume` unconditionally even if `used` is zero.
https://github.com/rust-lang-nursery/futures-rs/blob/f4dad6654b65be08bb0b59f2a274ad18d96e436a/futures-util/src/io/read_until.rs#L33-L43
https://github.com/rust-lang-nursery/futures-rs/blob/f4dad6654b65be08bb0b59f2a274ad18d96e436a/futures-util/src/try_stream/into_async_read.rs#L149

This can be solved by consuming it only when `used` is non-zero.

However, the current implementation is the same behavior as std.
https://github.com/rust-lang/rust/blob/55c48b4e823ad3518bb8e6d3fd7da4f2fc0c5fc2/src/libstd/io/mod.rs#L1475-L1492

Reply:
Sounds like `IntoAsyncRead` should special-case `consume(0)` to be a noop then.

Reply:
Sounds good. Opened #1584.